Source/JavaScriptCore/runtime/StableSort.h

```diff
@@ -45,54 +45,46 @@ static ALWAYS_INLINE bool coerceComparatorResultToBoolean(JSGlobalObject* global
 }
 
 template<typename ElementType, typename Functor>
-static ALWAYS_INLINE void arrayInsertionSort(VM& vm, std::span<ElementType> span, const Functor& comparator)
+static ALWAYS_INLINE void arrayInsertionSort(VM& vm, std::span<ElementType> span, const Functor& comparator, size_t sortedHeader = 0)
 {
     auto scope = DECLARE_THROW_SCOPE(vm);
 
     auto* array = span.data();
     size_t length = span.size();
-    for (size_t i = 1; i < length; ++i) {
+    for (size_t i = sortedHeader + 1; i < length; ++i) {
         auto value = array[i];
-        size_t j = i;
-        for (; j > 0; --j) {
-            auto target = array[j - 1];
+        // [l, r)
+        size_t left = 0;
+        size_t right = i;
+        for (; left < right;) {
+            size_t m = left + (right - left) / 2;
+            auto target = array[m];
             bool result = comparator(value, target);
             RETURN_IF_EXCEPTION_WITH_TRAPS_DEFERRED(scope, void());
             if (!result)
-                break;
-            array[j] = target;
+                left = m + 1;
+            else
+                right = m;
         }
-        array[j] = value;
+        ElementType t = value;
+        for (size_t j = left; j < i; ++j)
+            std::swap(array[j], t);
+        array[i] = t;
     }
 }
 
 template<typename ElementType, typename Functor>
-static ALWAYS_INLINE void arrayMerge(VM& vm, std::span<ElementType> dst, std::span<const ElementType> src, size_t srcIndex, size_t srcEnd, size_t width, const Functor& comparator)
+static ALWAYS_INLINE void mergePowersortRuns(VM& vm, std::span<ElementType> dst, std::span<const ElementType> src, size_t srcIndex1, size_t srcEnd1, size_t srcIndex2, size_t srcEnd2, const Functor& comparator)
 {
     auto scope = DECLARE_THROW_SCOPE(vm);
 
-    size_t left = srcIndex;
-    size_t leftEnd = std::min<size_t>(left + width, srcEnd);
-    size_t right = leftEnd;
-    size_t rightEnd = std::min<size_t>(right + width, srcEnd);
-
-    if (right >= rightEnd) {
```

Comment:
This part is very important optimization (without it, SP3 will regress). Is it ok to remove this?